### ISSUE TITLE
Add support for importing movary exports

### ISSUE THREAD

**Issue by sahinakkaya** [sbondCo/Watcharr#332]
It would be really nice to import [movary](https://github.com/leepeuker/movary) data into this app. I am currently using movary and it has support for exporting its data in csv. I want to give this app a try but without my data, I can't switch even if I like it more than movary.

**Comment by IRHM** [comment]
Hi @sahinakkaya, thanks for making an issue!

I hadn't heard of Movary before, but the csv importing should be simple to add alongside the TMDB importing (mostly telling it to look for different fields).

I'll see if I can add this in for you later tonight.

**Comment by sahinakkaya** [comment]
Thank you for your interest into making this possible! To make things easier for you, here are the 3 files exported from movary:

Watch history: [history.csv](https://github.com/sbondCo/Watcharr/files/14112569/history.csv)
User rating of the watched movies: [ratings.csv](https://github.com/sbondCo/Watcharr/files/14112571/ratings.csv)
User's watch list: [watchlist.csv](https://github.com/sbondCo/Watcharr/files/14112573/watchlist.csv)

I've exported them from my instance and simplified it for using it with basic testing scenarios.

**Comment by sahinakkaya** [comment]
Btw, there are probably edge cases where a movie can be watched multiple times and appear in watch history more than once. I don't know if Watcharr supports this so I didn't want to include the same movie in multiple date scenario. You might need to re adjust the files if Watcharr supports this feature too. (i. e. edit `history.csv` to add another entry for the same movie just with different time and optionally a comment)

**Comment by IRHM** [comment]
@sahinakkaya Just getting a chance to look into this now. Unfortunately Watcharr doesn't support adding multiple plays, I usually just add how many plays of a movie I have into the comment/thoughts section (If you'd like to see this feature, I can look into adding that).

Trying to figure out the difference between the export files. I'm guessing the watchlist.csv is your actual watchlist, then history.csv is how many plays you have on items in your watchlist, is that right?

If that is right, I think I can try combining all files into one before import so we have your watched list with the comment and rating saved.

**Comment by IRHM** [comment]
Added most of the logic for this to work I think, just going to wait for your reply to my last comment before doing the rest (time for 💤).

**Comment by sahinakkaya** [comment]
Watch list is the movies I want to watch in the future ("Planned" in watcharr). There are dates in watch list. Dates mean "when" I added that movie to my watch list.

History is the movies that I've watched ("Watched list" in watcharr). A movie can be watched multiple times, hence can appear in history multiple times.

Ratings is my ratings for the watched movies. There can be only one rating record for a movie no matter how many times it appears in history.

**Comment by sahinakkaya** [comment]
>Unfortunately Watcharr doesn't support adding multiple plays

I've experimented in Watcharr a little bit. Can't the following be used to represent multiple plays? Every time status is changed to finished would mean a movie is watched again. 

![image](https://github.com/sbondCo/Watcharr/assets/32161460/000eeb45-6ba9-4e01-9cbe-3f5a58dbc920)


 It is not a big deal if this is hard to implement. I can live without it. In that case, I think keeping the date of first time a movie appears in history would make more sense as I generally wonder when is the "first" time I've watched a movie.

**Comment by IRHM** [comment]
@sahinakkaya Thanks for the info, I should be able to finish it up now.

Interesting idea using the finished activity to count as a play, it could work well for movies, but with tv shows, I sometimes go back to watching if a new season comes out, then back to finished, but wouldn't want that to count as a play (though we could add a new `countAsPlay` property to activity so it can be changed manually).

I'll make a new ticket for that so we can think about it.

**Comment by IRHM** [comment]
@sahinakkaya I'll release a new version shortly for you to test tomorrow.

Your example files import fine for me and it works when adding some duplicate entries for movies in history.csv (if it finds dupes, it will get all the dates and add them as separate activities, you may need to refresh your browser if the activity dates are wrong, couldn't figure that bug out so it'll have to stay for now :().

The issue will close when the PR is merged, but please feel free to re-open or open a new issue if the spaghetti code doesn't work for you and we can try to figure it out.

**Comment by sahinakkaya** [comment]
> it could work well for movies, but with tv shows, I sometimes ...

If it works for movies then there is no problem. Because you can't track tv shows in movary.

Btw, I could have test it using the branch from PR. You had not need to merge and create a new release. Anyway, I will look test this soon. Thank you for implementing this!

**Comment by sahinakkaya** [comment]
I've tried to import movies from movary to test the this. Ratings seems to have correctly imported. There are some minor problems with planned and watched movies however. For example, here is a watched movie activity:
![image](https://github.com/sbondCo/Watcharr/assets/32161460/eb2d0a2b-7bf7-4f14-96f8-6c04d5a5577a)

"Imported watch date" is correct but "Imported" activity shouldn't be there because it messes up the sorting order of the movies. A movie should show up first if I watched it recently but it is showing me movies from 2008 first because it has more recent activity.

Second problem is some movies are missing. I sometimes watch a movie and add a play, but still keep that movie in my watch list to watch it again later. These movies are missing in "Watched movies" list. They are visible in "Planned movies" list. What I expect to see is this:

Status: Planned

Activity:
- Imported Planned Date
- Imported Watch Date

This second issue might be complicated to implement so I don't expect it to be fixed. I can live with it. I reported it to make sure we know about it. And I hope first problem is easy to fix. 

I can't open issue again so you might want to reopen it. @IRHM

**Comment by sahinakkaya** [comment]
Movies with multiple watch dates also works. Just that the last "Imported" activity shouldn't be there.

![image](https://github.com/sbondCo/Watcharr/assets/32161460/8a5c2e96-e7bb-4d21-94f4-5c8afc622612)

**Comment by IRHM** [comment]
Hi @sahinakkaya, thank you for the feedback.

> "Imported watch date" is correct but "Imported" activity shouldn't be there because it messes up the sorting order of the movies.

I see, I forget, but I believe the sorting on your watcharr watchlist is determined by a `updatedAt` on the actual watch entry in the database.

So I guess the problem is when importing a movie, we want to change the `updatedAt` (note for self: probably set `createdAt` to this date too, so if movie is updated, original createdAt sorting could still work) on the watchlist entry to the date set in your Movary history.

I think that should be doable, will have a look in a few hours if I can (or later tonight if i cant find the time).

---

On the second issue, I believe I could add a few extra special cases to get it working for you.

If you are able to upload the export csv files with this case for me to test properly, that would be great!


Thanks again for trying out the changes!

**Comment by sahinakkaya** [comment]
I can upload relevant test cases in a few hours. Thanks :)

**Comment by sahinakkaya** [comment]
>I can upload relevant test cases in a few hours.

I gave up from that feature because I think it is not possible to achieve it in Watcharr without changing its internal logic. Consider that I have watched "Oppenheimer". I mark this movie as "Watched" in Watcharr. After some time, I decided I will watch Oppenheimer again. I mark it as "Planned". Now, when I filter watched movies, I won't see Oppenheimer even though I watched it. This behaviour is something you should decide. If you think it is better this way, then movary data can't be imported with 100% accuracy. And that is OK for me too. It is not reasonable to expect 100% compatibility because these are 2 different apps. If you change your mind, however, I can share you required files to test it.

The first issue I mentioned in earlier comment is still relevant though. And it is more important to me.

**Comment by IRHM** [comment]
@sahinakkaya Oh, I see, I didn't realise that was how it may be used. You are right, a movie/show can only have one status.

The UI has all activity for your content, we could probably get around this by checking past activity for a Watched entry when filtering for watched (eg if oppenheimer has planned status, but has a watched/finished activity entry, we could display it when filtering for finished content).

If you are interested I can still look into it for you and implement it like that ^. Though it would probably be a user setting that controls how that behaves.



Will definitely get the first issue fixed for you.

**Comment by sahinakkaya** [comment]
Since I am just a user, and this is a work that you do voluntarily, I can't say I want it. It is up to you to implement it or not. I am totally fine with both decisions. And thanks for all the effort you put in this project. I really appreciate it ^^

**Comment by IRHM** [comment]
Hi @sahinakkaya, both problems should be fixed now. You can try it out by checking out the dev branch (default branch on this repo) or by using the latest [dev image](https://github.com/sbondCo/Watcharr/pkgs/container/watcharr/174392309?tag=dev) if you prefer.

For the first problem, we now set the `createdAt` and `updatedAt` fields to the date in your history, so the sorting should be working as intended now.

For the second problem, I've added a new setting `Include Previously Watched`, which you can toggle in your profile. This should change the 'Finished' filters behaviour to include your watched content that now has a different status.

Hopefully it works for you, let me know how it goes!

**Comment by sahinakkaya** [comment]
Both issues are now fixed as you mentioned. But now something else is broken. Watch dates were correctly imported here: https://github.com/sbondCo/Watcharr/issues/332#issuecomment-1923287334

Now, the situation is this for the same movie:
![image](https://github.com/sbondCo/Watcharr/assets/32161460/afb19a2a-f408-4359-b8c0-880e923b35c8)

I checked other movies, it is not just one record. Every one of them has the same watch date duplicated. 

Also, in the profile page, it would be a nice change to calculate statistics based on new option "Include Previously Watched". 

![image](https://github.com/sbondCo/Watcharr/assets/32161460/f6de76ea-de07-4501-a885-100430b7dd36)

The above statistics are calculated based on "Watched" movies. Since we have different meaning for what is considered as watched now, I think it would be nice if these numbers would be consistent with that option.

**Comment by IRHM** [comment]
Hi @sahinakkaya, thanks for trying it out and being patient with the bugs.

The first issue should hopefully fix itself if you refresh the page. There seems to be a bug, where multiple activities are shown on the same date in ui right after being imported, but in the database they are correct (hopefully for you too), so refreshing should sort it out.

I'll have a look if we can show previously watched in the stats too.

**Comment by sahinakkaya** [comment]
>The first issue should hopefully fix itself if you refresh the page

Yeah, that fixed it. Strange bug :) 

>I'll have a look if we can show previously watched in the stats too.

yay!

**Comment by sahinakkaya** [comment]
@IRHM I realized another issue. I can say I am a good tester :) I hope these issues are not overwhelming for you. 

When "Include Previously Watched" is enabled in settings, filtering by type (i.e. Movie, Show) doesn't have any effect. It always shows both movies and shows regardless of the selection. 
<img width=287 src="https://github.com/sbondCo/Watcharr/assets/32161460/7f553738-69dd-42b9-9032-010e13aa077e"/>

**Comment by IRHM** [comment]
Hey @sahinakkaya, you are a good tester, good catch!

I don't have time to properly test at the moment, but I think all the issues should be solved in the [awesome-stuff-too](https://github.com/sbondCo/Watcharr/tree/awesome-stuff-too) branch. I'll be back shortly to test it all, in the meantime you can check it out if you'd like.

Thanks again, I appreciate all the feedback and testing!

**Comment by sahinakkaya** [comment]
Hi @IRHM , I've tested `awesome-stuff-too` branch and I can confirm it is really awesome :-) All the issues we discussed seems to have been fixed. Writing down what I've tested to make sure we are on the same page:

- Filtering Movie/Show works as intended regardless of the "Include Previously Watched" option's status
- When "Include Previously Watched" is selected, statistics are updated correctly
- "Include Previously Watched" option is also correctly working on TV shows. 


Thank you very much for your efforts! I guess we can close this now.
